### ISSUE TITLE
Should Agent have a set method as well as AgentBuilder?

### ISSUE THREAD

**Issue by jsha** [algesten/ureq#203]
Pulling out a conversation from https://github.com/algesten/ureq/pull/201: 

> Keep set on Agent level as well as AgentBuilder.
>
> The idea is that an auth exchange might result in a header that need
to be set after the agent has been built.

I'd like to approach this use case a different way:

```rust
let builder = ureq::AgentBuilder::new()
  .set("Some-Header", "foo"));
let agent = builder.build();
let resp = agent.get("https://example.com/fetch-token")?;
let authed_agent = builder
  .set("Token", resp.header("Token").unwrap())
  .build();
let resp2 = agent.post("https://example.com/authed_endpoint")?;
```

In other words, make `AgentBuilder.build()` not consume self. Alternately, we could offer an `AgentBuilder.clone()`, but that is a little more awkward.

Here's my reasoning: We're intentionally making an Agent really easy to share across threads by cloning it. We can protect the Agent's internal state with a Mutex so we don't have memory-level race conditions. But if an Agent can have headers added after construction, we could have application-level race conditions: One thread might be performing an auth handshake while another is making a request, and that request would be authed or not depending on the random ordering of the two. By requiring the headers to be set at build time, we make it easier to reason about any given Agent: it's authed, or it's not authed, and that won't change.

I have to acknowledge, of course, that the above logic could equally well apply to auth cookies, but the way cookies work means we kind of need to allow mutation of an Agent's cookie jar during its lifetime.

Also, if we allow setting headers on an already-built Agent, we will probably eventually get requests to allow removing headers from an already-built Agent; it has the potential to get quite complicated. By keeping our API surface for setting headers to two places (AgentBuilder and RequestBuilder) we can hopefully keep things simple.

The main thing standing in the way of the code snippet I posted above is that `AgentBuilder.build()` _does_ consume the builder. That's because `CookieStore` is not cloneable. We could ask upstream about adding a clone or (my preference for now) we could manually implement a clone for our `CookieTin` wrapper by copying out all the cookies.

**Comment by algesten** [comment]
Not sure I entirely agree here. I admit I have put the `headers` vec in a bad place, however, `Agent` is absolutely a mutable shared state. Breaking the builder pattern + potentially requiring all conf in the builder to be clonable, sounds on balance worse to me.

Like, if we are reusing connections and have a mutable cookie store, why not some mutable headers?

As long as we make the contract clear: headers are copied into `Request` when the request is created.

**Comment by jsha** [comment]
> Breaking the builder pattern

It's funny, I think of `Agent.set` as somewhat breaking the builder pattern. My thinking is that, by separating out various configuration into the builder, we enforce that it is only set once, and separate the process of configuring from the process of using. So for instance the Agent doesn't have to have setters and getters for its various fields: the Builder has setters, and the Agent only has getters where appropriate. This also allows us to sidestep the question of calling methods `set_foo`, `get_foo`, or `foo` by letting everything be `foo`. :-)

> Agent is absolutely a mutable shared state.
> Like, if we are reusing connections and have a mutable cookie store, why not some mutable headers?

I agree that Agent is mutable shared state. It's a question of how much of it should be mutable. After headers, why not allow mutating the TLS config? Or the number of connections in the pool? Those would definitely an acceptable choice, and it's what we currently have in the 1.x series. But I think we get some really nice API clarity by saying that most things are set at build time.

Here's a concrete use case: For most binaries, it's ideal to create one Agent for the whole program, so you can maximally reuse connections for the whole program's lifetime. If you have a large program, and you are sharing out references to that one agent to every corner of it, how can you be confident you don't have spooky action at a distance, where component 1 modifies that agent to add a header, which affects calls that component 2 is making?

For most types, that problem is solved by mutability rules. Only mutable references can change state, so if you don't want the rest of your program to change your Foo, don't let any other components get an `&mut Foo`. Agent uses interior mutability, so any component with a clone of Agent can change state.

Anyhow, something to think about. I appreciate that we can talk through design questions like this together!

**Comment by algesten** [comment]
> If you have a large program, and you are sharing out references to that one agent to every corner of it, how can you be confident you don't have spooky action at a distance, where component 1 modifies that agent to add a header, which affects calls that component 2 is making?

Ah, ok. This sways me. In fact. Maybe headers on a Agent level is a non-starter, and we should just remove it altogether. And let me back that up by joining in the discussion from #209 

> In theory, since the user is calling set_cookie, we could trust them to set the Domain field to the domain they want the cookie stored on instead of the Domain field parsed out of the cookie header. But this would not allow the user to specify the "host only" behavior I described for when the domain is empty. It would also be risky, since a user that naively parsed a Set-Cookie header without checking it against the request host would allow evil.com to set cookies for example.com.

This is also an argument that convinces me. To elaborate, cookies already have a host separation built into the standard, so does our connection pooling. Setting headers on an agent level doesn't (at least not with the API we provide).

So I u-turn :) I think we should remove both Agent/AgentBuilder headers as well as `set_cookie`. 

In a future iteration we could consider a consistent mutate-state-per-host API on Agent. Something like.

```rust
impl Agent {
    pub fn mutate_host(&self, host: &str) -> HostMutate {
       ...
    }
}

pub struct HostMutate<'a> { … }

impl <'a> HostMutate<'a> {
  pub fn add_cookie(&mut self, cookie: Cookie) { … }
  pub fn add_header(&mut self, key: &str, val: &str) { … }
  pub fn close_pooled(&mut self) { … }
}
```

**Comment by algesten** [comment]
Closed by #213
